File: HyeonjunLee/BOJ/Gold/탑.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        Stack<Integer> stack = new Stack<>();
        Stack<Integer> idx = new Stack<>();

        int n = Integer.parseInt(br.readLine());
        int[] answer = new int[n];
        int[] arr = new int[n];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        stack.push(arr[arr.length - 1]);
        idx.push(arr.length - 1);
        for (int i = n - 2; i > -1; i--) {
            while (stack.peek() <= arr[i]) {
                answer[idx.peek()] = i + 1;

                stack.pop();
                idx.pop();

                if (stack.isEmpty()) break;
            }
            stack.push(arr[i]);
            idx.push(i);
        }

        for (int j : answer) {
            System.out.print(j + " ");
        }
    }
}

File: syJoh/BOJ/탑.py
import sys
input = sys.stdin.readline

n = int(input())
toplist = list(map(int, input().split()))
answer = [0]*n
stack = []

for i in range(len(toplist)):
    while stack: #stack이 비어있지 않은 동안 실행
        if stack[-1][1] > toplist[i]:
            answer[i] = stack[-1][0] + 1 #인덱스 1부터 시작
            break
        else:
            stack.pop()
    stack.append((i, toplist[i])) #현재 탑의 인덱스와 높이 추가
print(*answer)

"""
import java.util.*;
import java.io.*;

class Pair{
    int num;
    int idx;
    Pair(int num, int idx){
        this.num = num;
        this.idx = idx;
    }
}
public class Main{
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        Stack<Pair> stack = new Stack<>();

        for(int i=1; i<=n; i++){
            int num = Integer.parseInt(st.nextToken());
            while(!stack.isEmpty()){
                Pair cur = stack.peek();
                if(cur.num >= num){
                    System.out.print(cur.idx+" ");
                    break;
                }
                stack.pop();
            }
            if(stack.isEmpty()){
                System.out.print("0 ");
            }
            stack.add(new Pair(num, i));
        }
    }
}

"""
